File: ios/Classes/RicohThetaPlugin.m
#import "RicohThetaPlugin.h"
#import "HttpConnection.h"
#import "PictureController.h"
#import "StorageController.h"
#import "Constants.h"

FlutterEventSink livePreviewStreamEventSink;
FlutterEventSink downloadStreamEventSink;

PictureController *pictureController;
HttpConnection *httpConnection;
StorageController *storageController;

@implementation RicohThetaPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
                                   methodChannelWithName:@"ricoh_theta"
                                   binaryMessenger:[registrar messenger]];
  
  LivePreviewStreamHandler *livePreviewStreamHandler =
  [[LivePreviewStreamHandler alloc] init];
  FlutterEventChannel *livePreviewChannel = [FlutterEventChannel eventChannelWithName:@"ricoh_theta/preview"
                                                                      binaryMessenger:[registrar messenger]];
  [livePreviewChannel setStreamHandler:livePreviewStreamHandler];
  
  DownloadStreamHandler *downloadStreamHandler =
  [[DownloadStreamHandler alloc] init];
  FlutterEventChannel *downloadChannel = [FlutterEventChannel eventChannelWithName:@"ricoh_theta/download"
                                                                   binaryMessenger:[registrar messenger]];
  [downloadChannel setStreamHandler:downloadStreamHandler];
  
  RicohThetaPlugin *instance = [[RicohThetaPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
  
  httpConnection = [[HttpConnection alloc] init];
  pictureController = [[PictureController alloc] init];
  storageController = [[StorageController alloc] init];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  [self bindResultToControllers:result];
  
  if ([@"setTargetIp" isEqualToString:call.method]) {
    [self _handleSetTargetIp:call result:result];
  } else if ([@"disconnect" isEqualToString:call.method]) {
    [self _handleDisconnect:call result:result];
  } else if ([@"startLiveView" isEqualToString:call.method]) {
    [self _handleStartLiveView:call result:result];
  } else if ([@"resumeLiveView" isEqualToString:call.method]) {
    [self _handleResumeLiveView:call result:result];
  } else if ([@"pauseLiveView" isEqualToString:call.method]) {
    [self _handlePauseLiveView:call result:result];
  } else if ([@"removeImageWithFileId" isEqualToString:call.method]) {
    [self _handleRemoveImageWithFileId:call result:result];
  } else if ([@"stopLiveView" isEqualToString:call.method]) {
    [self _handleStopLiveView:call result:result];
  } else if ([@"adjustLiveViewFps" isEqualToString:call.method]) {
    [self _handleAdjustLiveViewFps:call result:result];
  } else if ([@"batteryLevel" isEqualToString:call.method]) {
    [self _handleBatteryLevel:call result:result];
  } else if ([@"getStorageInfo" isEqualToString:call.method]) {
    [self _handleStorageInfo:call result:result];
  } else if ([@"getImage" isEqualToString:call.method]) {
    [self _handleGetImageWithFileId:call result:result];
  } else if ([@"getImageInfoes" isEqualToString:call.method]) {
    [self _handleGetImageInfoes:call result:result];
  } else if ([@"getDeviceInfo" isEqualToString:call.method]) {
    [self _handleDeviceInfo:call result:result];
  } else if ([@"takePicture" isEqualToString:call.method]) {
    [self _handleTakePicture:call result:result];
  } else if ([@"update" isEqualToString:call.method]) {
    [self _handleUpdate:call result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)_handleUpdate:(FlutterMethodCall*)call result:(FlutterResult)result {
  [httpConnection update];
  result(nil);
}

- (void)_handleBatteryLevel:(FlutterMethodCall*)call result:(FlutterResult)result {
  [httpConnection getBatteryLevel:^(const NSNumber *battery) {
    if (!battery) {
      result([FlutterError errorWithCode:@"BATTERY_ERROR" message:@"unable to retrieve battery level" details:nil]);
      return;
    }
    
    result(battery);
  }];
}

- (void)_handleDisconnect:(FlutterMethodCall*)call result:(FlutterResult)result {
  [httpConnection close:^{
    result(nil);
  }];
}

- (void)_handleAdjustLiveViewFps:(FlutterMethodCall*)call result:(FlutterResult)result {
  float fps = [call.arguments[@"fps"] floatValue];
  
  if (!fps) {
    result([FlutterError errorWithCode:@"WRONG_FPS" message:@"fps value invalid must not be null" details:nil]);
  }
  
  [pictureController adjustLiveViewFps:fps];
}

- (void)_handleRemoveImageWithFileId:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *fileId = call.arguments[@"fileId"];
  
  if (!fileId) {
    result([FlutterError errorWithCode:@"MISSING_FILE_ID" message:@"file id need to be specified" details:nil]);
  }
  
  [storageController removeImageWithFileId:fileId];
}

- (void)_handleGetImageWithFileId:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *fileId = call.arguments[@"fileId"];
  NSString *path = call.arguments[@"path"];
  
  if (!fileId) {
    result([FlutterError errorWithCode:@"MISSING_FILE_ID" message:@"file id need to be specified" details:nil]);
  }
  
  if (!path) {
    result([FlutterError errorWithCode:@"MISSING_PATH" message:@"path need to be specified" details:nil]);
  }
  
  [storageController getImageWithFileId:fileId andPath:path];
}

- (void)_handleStartLiveView:(FlutterMethodCall*)call result:(FlutterResult)result {
  float fps = [call.arguments[@"fps"] floatValue];
  
  [pictureController startLiveView:fps];
}

- (void)_handleGetImageInfoes:(FlutterMethodCall*)call result:(FlutterResult)result {
  [storageController getImageInfoes];
}

- (void)_handleResumeLiveView:(FlutterMethodCall*)call result:(FlutterResult)result {
  [pictureController resumeLiveView];
}

- (void)_handlePauseLiveView:(FlutterMethodCall*)call result:(FlutterResult)result {
  [pictureController pauseLiveView];
}

- (void)_handleStopLiveView:(FlutterMethodCall*)call result:(FlutterResult)result {
  [pictureController stopLiveView];
}

- (void)_handleTakePicture:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *path = call.arguments[@"path"];
  
  if (!path) {
    result([FlutterError errorWithCode:@"MISSING_PATH" message:@"path need to be specified" details:nil]);
  }
  
  [pictureController takePictureWithPath:path];
}

- (void)_handleStorageInfo:(FlutterMethodCall*)call result:(FlutterResult)result {
  [storageController getStorageInfo];
}

- (void)_handleSetTargetIp:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *ipAddress = call.arguments[@"ipAddress"];
  
  if (!ipAddress) {
    result([FlutterError errorWithCode:@"MISSING_IP_ADDRESS" message:@"ip address need to be specified" details:nil]);
  }
  
  [httpConnection setTargetIp:ipAddress];
  [self bindHttpConnectionToControllers];
  result(nil);
}

- (void)bindResultToControllers:(FlutterResult)result {
  [pictureController setResult:result];
  [storageController setResult:result];
}

- (void)bindHttpConnectionToControllers {
  [pictureController setHttpConnection:httpConnection];
  [storageController setHttpConnection:httpConnection];
}

- (void)_handleDeviceInfo:(FlutterMethodCall*)call result:(FlutterResult)result {
  [httpConnection getDeviceInfo:^(const HttpDeviceInfo *info) {
    if (!info) {
      result([FlutterError errorWithCode:@"INFO_ERROR" message:@"unable to retrieve device info" details:nil]);
      return;
    }
    
    result(@{
      @"model": info.model,
      @"firmwareVersion": info.firmware_version,
      @"serialNumber": info.serial_number
    });
  }];
}

@end


@implementation LivePreviewStreamHandler
- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
  livePreviewStreamEventSink = eventSink;
  [pictureController setLivePreviewEventSink:eventSink];
  return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
  livePreviewStreamEventSink = nil;
  [pictureController setLivePreviewEventSink:livePreviewStreamEventSink];
  return nil;
}
@end

@implementation DownloadStreamHandler
- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
  downloadStreamEventSink = eventSink;
  [storageController setDownloadEventSink:eventSink];
  return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
  downloadStreamEventSink = nil;
  [storageController setDownloadEventSink:downloadStreamEventSink];
  return nil;
}
@end

File: android/src/main/java/com/apparence/ricoh_theta/RicohThetaPlugin.java
package com.apparence.ricoh_theta;

import android.os.StrictMode;

import androidx.annotation.NonNull;

import com.apparence.ricoh_theta.task.BatteryTask;
import com.apparence.ricoh_theta.task.DeviceInfoTask;
import com.apparence.ricoh_theta.task.StorageInfoTask;
import com.theta360.sdk.v2.network.HttpConnector;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** RicohThetaPlugin */
public class RicohThetaPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private HttpConnector camera;

  // Event channel
  private EventChannel livePreviewChannel;
  private EventChannel downloadChannel;

  // Controllers
  private PictureController pictureController = new PictureController();
  private StorageController storageController = new StorageController();

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "ricoh_theta");
    channel.setMethodCallHandler(this);

    livePreviewChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "ricoh_theta/preview");
    downloadChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "ricoh_theta/download");

    livePreviewChannel.setStreamHandler(pictureController);
    downloadChannel.setStreamHandler(storageController);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    // FIXME: Don't know if it's a good practice to disable this ?
    StrictMode.ThreadPolicy policy =
            new StrictMode.ThreadPolicy.Builder().permitAll().build();
    StrictMode.setThreadPolicy(policy);

    _bindResultToControllers(result);

    if (call.method.equals("setTargetIp")) {
      _handleSetTargetIp(call, result);
    } else if (call.method.equals("getDeviceInfo")) {
      _handleDeviceInfo(call, result);
    } else if (call.method.equals("takePicture")) {
      _handleTakePicture(call, result);
    } else if (call.method.equals("disconnect")) {
      _handleDisconnect(call, result);
    } else if (call.method.equals("startLiveView")) {
      _handleStartLiveView(call, result);
    } else if (call.method.equals("removeImageWithFileId")) {
      _handleRemoveImageWithFileId(call, result);
    } else if (call.method.equals("getImage")) {
      _handleGetImage(call, result);
    } else if (call.method.equals("pauseLiveView")) {
      _handlePauseLiveView(call, result);
    } else if (call.method.equals("stopLiveView")) {
      _handleStopLiveView(call, result);
    } else if (call.method.equals("resumeLiveView")) {
      _handleResumeLiveView(call, result);
    } else if (call.method.equals("batteryLevel")) {
      _handleBatteryLevel(call, result);
    } else if (call.method.equals("getStorageInfo")) {
      _handleStorageInfo(call, result);
    } else if (call.method.equals("update")) {
      _handleUpdate(call, result);
    } else if (call.method.equals("getImageInfoes")) {
      _handleGetImageInfoes(call, result);
    } else if (call.method.equals("adjustLiveViewFps")) {
      _handleAdjustLiveViewFps(call, result);
    } else {
      result.notImplemented();
      return;
    }
  }

  private void _bindResultToControllers(Result result) {
    pictureController.setResult(result);
    storageController.setResult(result);
  }

  private void _handleRemoveImageWithFileId(MethodCall call, Result result) {
    String fileId = call.argument("fileId");

    if (fileId == null) {
      result.error("MISSING_FILE_ID", "file id need to be specified", "");
      return;
    }

    storageController.removeImageWithFileId(fileId);
  }

  private void _handleGetImage(MethodCall call, Result result) {
    String fileId = call.argument("fileId");
    String path = call.argument("path");

    if (fileId == null) {
      result.error("MISSING_FILE_ID", "file id need to be specified", "");
      return;
    }

    if (path == null) {
      result.error("MISSING_PATH", "path need to be specified", "");
      return;
    }

    storageController.getImageWithFileId(fileId, path);
  }

  private void _handleAdjustLiveViewFps(MethodCall call, Result result) {
    final Integer fps = call.argument("fps");

    if (fps == null) {
      result.error("MISSING_FPS", "fps need to be specified", "");
      return;
    } else {
      pictureController.setCurrentFps(fps);
      // FIXME: Improve this to not restart the preview, this is too heavy !
      pictureController.resumeLiveView();
    }
  }

  private void _handleGetImageInfoes(MethodCall call, Result result) {
    storageController.getImageInfoes();
  }

  private void _handleUpdate(MethodCall call, Result result) {
    // TODO
  }

  private void _handleResumeLiveView(MethodCall call, Result result) {
    pictureController.resumeLiveView();
  }

  private void _handleStopLiveView(MethodCall call, Result result) {
    pictureController.stopLiveView();
  }

  private void _handlePauseLiveView(MethodCall call, Result result) {
    // TODO
  }

  private void _handleStartLiveView(MethodCall call, Result result) {
    final Integer fps = call.argument("fps");

    if (fps == null) {
      result.error("MISSING_FPS", "fps need to be specified", "");
      return;
    } else {
      pictureController.startLiveView(fps);
    }

  }

  private void _handleDisconnect(MethodCall call, Result result) {
    // TODO
  }

  private void _handleTakePicture(MethodCall call, MethodChannel.Result result) {
    final String path = call.argument("path");

    pictureController.takePicture(path, result);
  }

  private void _handleStorageInfo(MethodCall call, Result result) {
    final StorageInfoTask storageInfoTask = new StorageInfoTask(camera, result);
    storageInfoTask.execute();
  }

  private void _handleBatteryLevel(MethodCall call, Result result) {
    final BatteryTask batteryTask = new BatteryTask(camera, result);
    batteryTask.execute();
  }

  private void _handleSetTargetIp(MethodCall call, Result result) {
    String ipAddress = call.argument("ipAddress");

    if (ipAddress == null) {
      result.error("MISSING_IP_ADDRESS", "ip address need to be specified", "");
      return;
    }

    camera = new HttpConnector(ipAddress);

    pictureController.setCamera(camera);
    storageController.setCamera(camera);

    result.success(null);
    return;
  }

  private void _handleDeviceInfo(MethodCall call, Result result) {
      final DeviceInfoTask deviceInfoTask = new DeviceInfoTask(camera, result);
      deviceInfoTask.execute();
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
